### DISCUSSION TITLE
DHCP assigned DNS servers

### DISCUSSION THREAD

**Discussion by hgschmie** [toitlang/toit#1333] [Q&A]
I am using the mqtt library and it could not connect to my hostname. This came as a bit of a surprise; I assumed that I could just do a `transport := mqtt.TcpTransport network --host=my.mqtt-host.my-domain` and be in business.

Browsing through the code, it seems that the toit runtime ignores the DHCP assigned DNS servers and always goes straight to 8.8.8.8. And that is even hardcoded: `lib/net/impl.toit` calls `dns.dns_lookup host` without a server so that falls back to the default.

I could resolve hostnames myself with `dns.dns_lookup` and pass in an IP of my local DNS server. Some problems:

- how do I find my local DNS server IPs? those have been handed in by DHCP, but how do find those?
- `dns.dns_lookup` only takes a single server. I have two for redundancy reasons. Short of building round robin myself, is there a way to pass in an array of DNS servers (I guess not).

Right now I am looking at hardcoding the IP of my local mqtt server in config. Not ideal.

**Comment by erikcorry** [comment]
Thanks for the bug report.

Certainly this can be done better.  I'll be landing some fixes that address your issues.

**Comment by hgschmie** [comment]
Thank you. Anything I can do to help? Happy to try things out.

**Comment by erikcorry** [comment]
First step: https://github.com/toitlang/toit/pull/1334

**Comment by erikcorry** [comment]
If there are multiple DNS servers would you prefer it to pick a random one, or is it better to start at the top?

**Reply by hgschmie** [reply]
not sure. According to docs, the linux resolver tries the hosts in order, one at a time and moves to the next if a request times out. So I guess that might be the best way to do that.

**Comment by erikcorry** [comment]
I didn't get to fixing this yet, but there is a workaround, so you can hard code your local DNS server.

I still plan to get this from the DHCP, so it's automatic.

For example to use the local DNS first and then fall back to Google:

```
import net.modules.dns

main:
  dns.default_client = dns.DnsClient ["192.168.0.1", "8.8.8.8"]
```

Beware that this will switch to the Google DNS if there is heavy congestion on the local network and a DNS request times out.  So you may want to just have just the one DNS server.

**Comment by erikcorry** [comment]
I changed it so it starts with the first, rather than using randomness to pick.

You have to be on the very newest release - I'm 99% sure this feature is in the latest SDK released in the new Jaguar at the end of last week.

**Comment by hgschmie** [comment]
yes, I upgraded to the latest SDK and this works for me. (setting DNS once at the beginning of my code, then just using the standard resolver. Thank you!

**Reply by kasperl** [reply]
@hgschmie SDK v2.0.0-alpha.53 uses DHCP to find the DNS server. Give it a try! It ships as part of Jaguar v1.9.1.

**Comment by hgschmie** [comment]
awesome. I gave it a try today. Here is what I see:

```
E (585) psram: No PSRAM detected
E (585) spiram: SPI RAM supported, but not found.
[toit] INFO: starting <v2.0.0-alpha.53>
[toit] DEBUG: clearing RTC memory: invalid checksum
[jaguar] INFO: container 'light-monitor' started
[wifi] DEBUG: connecting
[wifi] DEBUG: connected
[wifi] INFO: network address dynamically assigned through dhcp {ip: 192.168.2.192}
[wifi] INFO: DNS server address dynamically assigned through dhcp {dns: [192.168.2.18, 192.168.2.19]}
[light-monitor] DEBUG: Network established
[jaguar] INFO: running Jaguar device 'demo-lights' (id: '16839cf1-e76e-4568-9ae4-2ce70a6a8f5b') on 'http://192.168.2.192:9000'
[light-monitor] ERROR: Could not resolve host name: internal-time-host.domain!

******************************************************************************
Decoding by `jag`, device has version <2.0.0-alpha.53>
******************************************************************************
EXCEPTION error.
DNS lookup exception Server responded: NO_SUCH_DOMAIN```
```

Sadface. I added this to the DnsClient class (franky that should be there anyway):

```
--- ../toit/lib/net/modules/dns.toit	2023-02-08 18:08:46
+++ /Users/henning/.cache/jaguar/sdk/lib/net/modules/dns.toit	2023-02-08 18:33:35
@@ -161,6 +161,9 @@

   static DNS_UDP_PORT ::= 53

+  servers -> List:
+    return servers_
+
   /**
   Look up a domain name and return an A or AAAA record.
```

and logged the result out at startup:

```
    LOG.debug "Assigned DNS servers: $dns.default_client.servers"
```

```
[wifi] INFO: network address dynamically assigned through dhcp {ip: 192.168.2.192}
[wifi] INFO: DNS server address dynamically assigned through dhcp {dns: [192.168.2.18, 192.168.2.19]}
[light-monitor] DEBUG: Network established
[jaguar] INFO: running Jaguar device 'demo-lights' (id: '16839cf1-e76e-4568-9ae4-2ce70a6a8f5b') on 'http://192.168.2.192:9000'
[light-monitor] DEBUG: Assigned DNS servers: [8.8.8.8, 1.1.1.1, 8.8.4.4, 1.0.0.1]
```

So yeah. It gets the right DNS from DHCP. But then, for whatever reason, the default DNS server is not the one that comes from DHCP but the default one internally.

When I poke some more at dns.toit and add

```
get_dhcp_client -> DnsClient?:
  return dhcp_client_
```

then it turns out that calling `dns.get_dhcp_client` returns null. So while I see the code in system/extensions/esp32/wifi.toit being executed, somehow I don't get the `dhcp_client_` field set.

(Yes, my device reports "FreeTOS" as platform, so  

So I built an SDK locally and flashed a patched firmware onto the device that basically does the same call  `logger_.info "Sanity check: $dns_module.get_dhcp_client.servers` right after the regular logging of the dns servers.

And lo and behold:

```
[wifi] DEBUG: connected
[wifi] INFO: network address dynamically assigned through dhcp {ip: 192.168.2.192}
[wifi] INFO: DNS server address dynamically assigned through dhcp {dns: [192.168.2.18, 192.168.2.19]}
[wifi] INFO: Sanity check: [192.168.2.18, 192.168.2.19]
[light-monitor] DEBUG: Network established
[jaguar] INFO: running Jaguar device 'device-lights' (id: '5a205a32-7c72-4619-8f7d-027475df4ad2') on 'http://192.168.2.192:9000'
[light-monitor] DEBUG: Default client servers: [8.8.8.8, 1.1.1.1, 8.8.4.4, 1.0.0.1]
[light-monitor] DEBUG: DHCP Client: null
```

so now I am stuck. The code definitely assigns a working DNS resolver to `dns..dhcp_client_`. I can read it back and it has the right DNS resolvers. But when my code starts up, the field is null and I get the default resolver.  *Something* is working but it is not propagating all the way down to the user code. 

Happy to try another version, for now I am back to hardcoded DNS.

**Comment by hgschmie** [comment]
any updates on this? I am comfortable with the state that my app is in right now, but I feel that this should be addressed at some point.

**Reply by floitsch** [reply]
@erikcorry was on vacation. He is back this week and will follow up.

**Comment by hgschmie** [comment]
Thanks for letting me know. @erikcorry  hope you had a great time away from work and computers. :-)

**Comment by erikcorry** [comment]
Unfortunately KLM has cancelled multiple flights, so I am not back home yet.

**Comment by hgschmie** [comment]
I assumed that when you wrote "KLM". ;-) (sorry, too many bad memories of cancelled flights from and to Schiphol). Only Air France is worse.
